homeassistant/components/nanoleaf/config_flow.py

```diff
@@ -0,0 +1,220 @@
+                step_id="user",
+                data_schema=USER_SCHEMA,
+                last_step=False,
+                errors={"base": "unknown"},
+            )
+        return self.async_show_form(step_id="link")
+
+    async def async_step_zeroconf(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf Zeroconf discovery."""
+        _LOGGER.debug("Zeroconf discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def async_step_homekit(self, discovery_info: DiscoveryInfoType) -> FlowResult:
+        """Handle Nanoleaf Homekit discovery."""
+        _LOGGER.debug("Homekit discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def _async_discovery_handler(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf discovery."""
+        host = discovery_info["host"]
+        # The name is unique and printed on the device and cannot be changed.
+        name = discovery_info["name"].replace(f".{discovery_info['type']}", "")
+        await self.async_set_unique_id(name)
+        self._abort_if_unique_id_configured({CONF_HOST: host})
+        self.nanoleaf = Nanoleaf(host)
+
+        # Import from discovery integration
+        self.device_id = discovery_info["properties"]["id"]
+        self.discovery_conf = cast(
+            dict,
+            await self.hass.async_add_executor_job(
+                load_json, self.hass.config.path(CONFIG_FILE)
+            ),
+        )
+        self.nanoleaf.token = self.discovery_conf.get(self.device_id, {}).get(
+            "token",  # >= 2021.4
+            self.discovery_conf.get(host, {}).get("token"),  # < 2021.4
+        )
+        if self.nanoleaf.token is not None:
+            self.context["source"] = config_entries.SOURCE_INTEGRATION_DISCOVERY
+            _LOGGER.warning(
+                "Importing Nanoleaf %s from the discovery integration", name
+            )
+            return await self.async_setup_finish()
+
+        self.context["title_placeholders"] = {"name": name}
+        return self.async_show_form(step_id="link")
+
+    async def async_step_link(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle Nanoleaf link step."""
+        if user_input is None:
+            return self.async_show_form(step_id="link")
+
+        try:
+            await self.hass.async_add_executor_job(self.nanoleaf.authorize)
+        except NotAuthorizingNewTokens:
+            return self.async_show_form(
+                step_id="link", errors={"base": "not_allowing_new_tokens"}
+            )
+        except Unavailable:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=USER_SCHEMA,
+                errors={"base": "cannot_connect"},
+                last_step=False,
+            )
```

Comment:
I'm pretty sure it landed here and it was't ready to pair. We should probably tell them to try to get it back ready to pair again and try again intead

Reply:
I've never experienced this. Should I send them back to the user step or let them try again?

Reply:
This was changed after another comment https://github.com/home-assistant/core/pull/52199#discussion_r694846182.
The flow is now aborted.